dango/types/src/vesting.rs

```diff
@@ -0,0 +1,183 @@
+        &self,
+        now: Timestamp,
+        vesting_amount: Uint128,
+    ) -> anyhow::Result<Uint128> {
+        if self.start_time + self.cliff > now {
+            return Ok(Uint128::ZERO);
+        }
+
+        let claim_percent = if now >= self.start_time + self.vesting {
+            Udec128::ONE
+        } else {
+            Udec128::checked_from_ratio(
+                (now - self.start_time).into_nanos(),
+                self.vesting.into_nanos(),
+            )?
+        };
+
+        Ok(vesting_amount.checked_mul_dec_floor(claim_percent)?)
+    }
+}
+
+#[grug::derive(Serde, Borsh)]
+pub struct Position<C = Undefined> {
+    pub user: Addr,
+    pub vesting_status: VestingStatus,
+    pub vested_token: Coin,
+    pub claimed_amount: Uint128,
+    pub claimable_amount: C,
+}
+
+impl Position {
+    pub fn new(user: Addr, vesting_schedule: Schedule, amount: Coin) -> Self {
+        Self {
+            user,
+            vesting_status: VestingStatus::Active(vesting_schedule),
+            vested_token: amount,
+            claimed_amount: Uint128::ZERO,
+            claimable_amount: Undefined::new(),
+        }
+    }
+
+    pub fn with_claimable_amount(
+        self,
+        now: Timestamp,
+        unlocking_schedule: &Schedule,
+    ) -> Position<Uint128> {
+        let claimable_amount = self
+            .compute_claimable_amount(now, unlocking_schedule)
+            .unwrap_or_default();
+
+        Position {
+            user: self.user,
+            vesting_status: self.vesting_status,
+            vested_token: self.vested_token,
+            claimed_amount: self.claimed_amount,
+            claimable_amount,
+        }
+    }
+}
+
+impl<T> Position<T> {
+    pub fn compute_claimable_amount(
+        &self,
+        now: Timestamp,
+        unlocking_schedule: &Schedule,
+    ) -> anyhow::Result<Uint128> {
+        // The claimable amount is the minimum between the claimable amount
+        // from the vesting status and the unlocking schedule
+        let claimable_amount = min(
+            self.vesting_status
+                .compute_claimable_amount(now, self.vested_token.amount)?,
+            unlocking_schedule.compute_claimable_amount(now, self.vested_token.amount)?,
+        );
+
+        Ok(claimable_amount
+            .checked_sub(self.claimed_amount)
+            .unwrap_or_default())
+    }
+
+    pub fn full_claimed(&self) -> bool {
+        match &self.vesting_status {
+            VestingStatus::Active(_) => self.vested_token.amount == self.claimed_amount,
+            VestingStatus::Terminated(terminated_amount) => {
+                *terminated_amount == self.claimed_amount
+            },
+        }
+    }
+}
+
+#[grug::derive(Serde, Borsh)]
+pub enum VestingStatus {
+    // Position is active. When active, the token stil being distributed
+    Active(Schedule),
+    // Position is terminated.
+    // When terminated, the snapshot of the vested so far is taken and stored
+    Terminated(Uint128),
```

Comment:
Docs should start with a triple slash (`///`) and with proper punctuation at the end (with period `.`).